File: src/Layout/Widgets/controller/io/Output.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package Layout.Widgets.controller.io;

import java.util.Objects;
import java.util.UUID;

/**
 *
 * @author Plutonium_
 */
public class Output<T> extends Put<T> {
    public final Id id;
    
        
    public Output(UUID id, String name, Class<T> c) {// System.out.println(id);
        super(c, null);
        this.id = new Id(id, name);
    }
    
    
    public String getName() {
        return id.name;
    }

    @Override
    public boolean equals(Object o) {
        if(this==o) return true;
        return o instanceof Output ? id.equals(((Output)o).id) : false;
    }

    @Override
    public int hashCode() {
        return 5 * 89 + Objects.hashCode(this.id);
    }
    
    public static class Id {
        public final UUID carrier_id;
        public final String name;

        public Id(UUID carrier_id, String name) {
            this.carrier_id = carrier_id;
            this.name = name;
        }

        @Override
        public boolean equals(Object o) {
            if(this==o) return true;
            if(o instanceof Id) 
                return ((Id)o).name.equals(name) && ((Id)o).carrier_id.equals(carrier_id);
            return false;
        }

        @Override
        public int hashCode() {
            int hash = 3;
            hash = 79 * hash + Objects.hashCode(this.carrier_id);
            hash = 79 * hash + Objects.hashCode(this.name);
            return hash;
        }

        @Override
        public String toString() {
            return name + "," + carrier_id.toString();
        }
        
        public static Id fromString(String s) {
            int i = s.indexOf(",");
            String n = s.substring(0,i);
            UUID u = UUID.fromString(s.substring(i+1, s.length()));
            return new Id(u,n);
        }
        
    }

}

File: src/player/main/sp/it/pl/layout/controller/io/Output.kt
package sp.it.pl.layout.controller.io

import java.util.Objects
import java.util.UUID
import sp.it.util.collections.materialize
import sp.it.util.type.VType

class Output<T>: Put<T> {
   val id: Id

   constructor(ownerId: UUID, name: String, type: VType<T>, initialValue: T): super(type, name, initialValue) {
      this.id = Id(ownerId, name)
   }

   /** Calls [sp.it.pl.layout.controller.io.Input.bind] on specified input with this output. */
   fun bind(input: Input<in T>) = input.bind(this)

   /** Calls [sp.it.pl.layout.controller.io.Input.unbind] on specified input with this output. */
   fun unbind(input: Input<*>) = input.unbind(this)

   /** @return true iff this output is bound to any [Input]. */
   fun isBound(): Boolean {
      return IOLayer.allLinks.keys.asSequence()
         .filter { it.key1()==this || it.key2()==this }
         .flatMap { sequenceOf(it.key1(), it.key2()).mapNotNull { it as? Input<*> } }
         .count()>0
   }

   /** Calls [sp.it.pl.layout.controller.io.Input.unbind] on all inputs bound to this output. */
   fun unbindAll() {
      IOLayer.allLinks.keys.asSequence()
         .filter { it.key1()==this || it.key2()==this }
         .flatMap { sequenceOf(it.key1(), it.key2()).mapNotNull { it as? Input<*> } }
         .materialize()
         .forEach { it.unbind(this) }
   }

   override fun equals(other: Any?) = this===other || other is Output<*> && id==other.id

   override fun hashCode() = 5*89 + Objects.hashCode(this.id)

   data class Id(val ownerId: UUID, val name: String) {

      override fun toString() = "$name,$ownerId"

      companion object {
         fun fromString(s: String): Id {
            val i = s.lastIndexOf(",")
            val n = s.substring(0, i)
            val u = UUID.fromString(s.substring(i + 1, s.length))
            return Id(u, n)
         }
      }

   }

}